source/slang/slang-emit-wgsl.cpp

```diff
@@ -0,0 +1,1013 @@
+}
+
+void WGSLSourceEmitter::emitTransposedMatrixType(
+    IRType *const elementType, IRMatrixType *const matrixType
+    )
+{
+    IRIntegerValue const colCountSlang {getIntVal(matrixType->getColumnCount())};
+    IRIntegerValue const rowCountSlang {getIntVal(matrixType->getRowCount())};
+    IRIntegerValue const colCountWGSL {rowCountSlang};
+    IRIntegerValue const rowCountWGSL {colCountSlang};
+    emitMatrixType(elementType, rowCountWGSL, colCountWGSL);
+}
+
+void WGSLSourceEmitter::emitMatrixType(
+    IRType *const elementType, IRIntegerValue const& rowCountWGSL,
+    IRIntegerValue const& colCountWGSL
+    )
+{
+    // WGSL uses CxR convention
+    m_writer->emit("mat");
+    m_writer->emit(colCountWGSL);
+    m_writer->emit("x");
+    m_writer->emit(rowCountWGSL);
+    m_writer->emit("<");
+    emitType(elementType);
+    m_writer->emit(">");
+}
+
+void WGSLSourceEmitter::emitStructDeclarationSeparatorImpl()
+{
+    m_writer->emit(",");
+}
+
+static bool isPowerOf2(uint32_t const n)
+{
+    return (n != 0U) && ((n - 1U) & n) == 0U;
+}
+
+void WGSLSourceEmitter::emitStructFieldAttributes(
+    IRStructType * structType, IRStructField * field
+    )
+{
+    // Tint emits errors unless we explicitly spell out the layout in some cases, so emit
+    // offset and align attribtues for all fields.
+    IRSizeAndAlignmentDecoration *const sizeAndAlignmentDecoration {
+        structType->findDecoration<IRSizeAndAlignmentDecoration>()
+    };
+    // NullDifferential struct doesn't have size and alignment decoration
+    if (sizeAndAlignmentDecoration == nullptr)
+        return;
+    SLANG_ASSERT(sizeAndAlignmentDecoration->getAlignment() > IRIntegerValue{0});
+    SLANG_ASSERT(
+        sizeAndAlignmentDecoration->getAlignment() <= IRIntegerValue{UINT32_MAX}
+    );
+    uint32_t const structAlignment {
+        static_cast<uint32_t>(sizeAndAlignmentDecoration->getAlignment())
+    };
+    IROffsetDecoration *const fieldOffsetDecoration {
+        field->findDecoration<IROffsetDecoration>()
+    };
+    SLANG_ASSERT(fieldOffsetDecoration->getOffset() >= IRIntegerValue{0});
+    SLANG_ASSERT(fieldOffsetDecoration->getOffset() <= IRIntegerValue{UINT32_MAX});
+    SLANG_ASSERT(isPowerOf2(structAlignment));
+    uint32_t const fieldOffset {
+        static_cast<uint32_t>(fieldOffsetDecoration->getOffset())
+    };
+    // Alignment is GCD(fieldOffset, structAlignment)
+    // TODO: Use builtin/intrinsic (e.g. __builtin_ffs)
+    uint32_t fieldAlignment = 1U;
+    while (((fieldAlignment & (structAlignment | fieldOffset)) == 0U))
+        fieldAlignment = fieldAlignment << 1U;
+
+    m_writer->emit("@align(");
+    m_writer->emit(fieldAlignment);
+    m_writer->emit(")");
+}
+
+bool WGSLSourceEmitter::isPointerSyntaxRequiredImpl(IRInst* inst)
+{
+    // Structured buffers are mapped to 'array' types, which don't need dereferencing
+    if (inst->getOp() == kIROp_RWStructuredBufferGetElementPtr)
```

Comment:
I wonder why is this needed.

Reply:
I added some examples [here](https://github.com/shader-slang/slang/pull/5006#discussion_r1744880039).